cloud,ccl,*: break `sql` -> `cloud/externalconn/utils` dependency

ZhouXing19 opened the pull request:
This commit is to break `cloud/externalconn/utils/`'s dependency on the `sql` package, as `sql` is just used to get the `sql.ExecCfg` for the validation function for external connection uri. But in fact, only a few fields of `sql.ExecCfg` are used; hence it makes more sense just to use a lighter `externalconn.ExternalConnEnv`.

Release note: None
Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96484)
<!-- Reviewable:end -->

ZhouXing19 commented:
nvm, I just realized maybe I should use `cloud.ExternalStorageContext`. Sorry for the noise -- changing it back to the draft mode.

ZhouXing19 commented:
I think it should be fine to not reuse `cloud.ExternalStorageContext`. The root of this dependency, `utils.CheckExternalStorageConnection`, will only be called when an external connection is created. It will trigger `ExternalStorageFromURIFactory` to create a new external storage, but the reverse doesn't happen. So I think it doesn't make sense to hook the fields needed for connection verification to `cloud.ExternalStorageContext`, and thus I made a new `externalconn.ExternalConnEnv` structure.

adityamaru commented on pkg/cloud/externalconn/utils/connection_utils.go:
```diff
@@ -32,11 +28,8 @@ const markerFile = "crdb_external_storage_location"
-) error {
-	cfg := execCfg.(*sql.ExecutorConfig)
-	es, err := cfg.DistSQLSrv.ExternalStorageFromURI(ctx, uri, user)
+func CheckExternalStorageConnection(ctx context.Context, env cloud.KMSEnv, uri string) error {
```
Why can't we pass in a `ExternalConnEnv` here instead of a cloud.KMSEnv? It seems a little strange to me to be passing around a KMS environment in a method that validates non-KMS connection URIs. From what I can tell passing in a `ExternalConnEnv` should give us a handle to all the things we need?

adityamaru commented on pkg/cloud/externalconn/impl_registry.go:
```diff
@@ -46,25 +50,27 @@ type schemeRegistration struct {
 	for _, v := range requestedValidations {
-		if validationFn, ok := s.validations[v]; ok {
-			err = errors.CombineErrors(err, validationFn(ctx, execCfg, user, uri.String()))
+		if validationFn, ok := s.validationsWithCloud[v]; ok {
+			err = errors.CombineErrors(err, validationFn(ctx, cloud.KMSEnv(&env), uri.String()))
+		}
+		if validationFn, ok := s.validationsWithSinkURI[v]; ok {
+			err = errors.CombineErrors(err, validationFn(ctx, env, uri.String()))
```
similar to my other comment I like the pattern of passing in an ExternalConnEnv to every validation method. It is then up to the validation methods to use that env to construct any validation-specific state. If we do that, I think we can revert this diff and go back to having a single validations map?

adityamaru commented on pkg/cloud/externalconn/impl_registry.go:
```diff
@@ -138,11 +164,73 @@ func ExternalConnectionFromURI(
+}
+
+// ValidationFnWithCloud is the type taken by Register*ValidationWithCloud. Validation functions can ping the external connection.
+type ValidationFnWithCloud func(ctx context.Context, env cloud.KMSEnv, uri string) error
```
Sorry for repeting myself thrice but just for clarity I think all validation functions should take in an ExternalConnEnv. It is then up to them to construct internal state required for the actual validation.

adityamaru commented on pkg/cloud/externalconn/impl_registry.go:
```diff
@@ -160,4 +248,14 @@ type TestingKnobs struct {
 func (t *TestingKnobs) ModuleTestingKnobs() {}
 
+// GetSkipCheckingExternalStorageConnection implements cloud.ExternalConnTestingKnobs.
+func (t *TestingKnobs) GetSkipCheckingExternalStorageConnection() func() bool {
```
personally I think this is a little less readable than just accessing the TestingKnob bools directly. I might be missing something but I'd prefer if our if conditions looked like:

`if knobs.SkipCheckingExternalStorageConnection != nil && knobs.SkipCheckingExternalConnection()`

instead of

`if knobs.GetSkipCheckingExternalStorageConnection() != nil && knobs.GetSkipCheckingExternalStorageConnection()()`

adityamaru reviewed (changes requested):
Thanks for doing this! I have a few high-level comments that I'm happy to discuss offline in more detail. If they make sense I think you can avoid touching anything in backupccl?

ZhouXing19 commented:
Per an offline convo with @adityamaru and we agreed on passing `externalconn.ExternalConnEnv` rather than `cloud.KMSCloud` to the validation functions. Have updated the PR and it's ready for another look.

adityamaru commented:
> Have updated the PR and it's ready for another look.

I think all the diffs in `backupccl` are formatting changes, mind reverting those so the number of files we touch is a lot less?

adityamaru commented on pkg/cloud/externalconn/impl_registry.go:
```diff
@@ -138,11 +138,71 @@ func ExternalConnectionFromURI(
 
-// ValidationFn is the type taken by Register*Validation. Validation functions can ping the external connection.
-type ValidationFn func(ctx context.Context, execCfg interface{}, user username.SQLUsername, uri string) error
+type ExternalConnEnv struct {
```
nit: you'll probably need a comment for this exported type and all the methods below. We used to have a linter that failed if you didn't.

ZhouXing19 replied:
Done.

adityamaru commented on pkg/cloud/externalconn/impl_registry.go:
```diff
@@ -138,11 +138,71 @@ func ExternalConnectionFromURI(
+
+	SkipCheckingExternalStorageConnection bool
+	SkipCheckingKMSConnection             bool
+	// ServerCfg is used to validate the external connection sink URI.
+	//// It implements *execinfra.ServerConfig.
+	ServerCfg interface{}
+	//ServerCfg *execinfra.ServerConfig
```
nit: Leftover commented code, and some cleanup. Can you maybe also add a line saying this is of type interface so as to not pull in a dependency to `execinfra`.

ZhouXing19 replied:
Done.

adityamaru commented on pkg/cloud/externalconn/impl_registry.go:
```diff
@@ -138,11 +138,71 @@ func ExternalConnectionFromURI(
+	externalStorageFromURIFactory cloud.ExternalStorageFromURIFactory,
+	skipCheckingExternalStorageConnection bool,
+	skipCheckingKMSConnection bool,
+	serverCfg *execinfra.ServerConfig,
```
should this be interface{} otherwise we're still pulling in `execinfra`?

ZhouXing19 replied:
Done.

adityamaru commented on pkg/sql/create_external_connection.go:
```diff
@@ -103,8 +103,27 @@ func (p *planner) createExternalConnection(
+	var SkipCheckingExternalStorageConnection bool
+	var SkipCheckingKMSConnection bool
+	if params.ExecCfg().ExternalConnectionTestingKnobs != nil {
+		SkipCheckingExternalStorageConnection = params.ExecCfg().ExternalConnectionTestingKnobs.SkipCheckingExternalStorageConnection()
```
don't we also have to check that `SkipCheckingExternal... != nil`?

ZhouXing19 replied:
Done

adityamaru reviewed (approved):
LGTM, just a couple of nits.

ZhouXing19 commented:
TFTR!
bors r=adityamaru

ZhouXing19 commented:
bors r=adityamaru

ZhouXing19 commented:
bors r=adityamaru

irfansharif commented:
bors r-

(I think this is missing some generated files `dev` expects.)

ZhouXing19 commented:
> (I think this is missing some generated files dev expects.)

Ran `./dev gen` locally and no new file changes 🤔

irfansharif commented:
Hm, I'm not sure, I see this above in the merge commit CI failure above:

<img width="533" alt="image" src="https://user-images.githubusercontent.com/10536690/217664631-1b69c0b2-9c41-4643-8d28-673fe0cc514f.png">

irfansharif commented:
I also see this in a commit CI from earlier:

<img width="1320" alt="image" src="https://user-images.githubusercontent.com/10536690/217666042-a8378158-9efb-4e3a-8c47-1bdb4ff49a88.png">

ZhouXing19 committed:
cloud,ccl,*: break sql -> cloud/externalconn/utils dependency

This commit is to break `cloud/externalconn/utils/`'s dependency of the `sql`
package, as `sql` is just used to get the `sql.ExecCfg` for the validation
function for external connection uri. But in fact, only a few fields of `sql.ExecCfg`
is used, hence it makes more sense to just use a lighter
`exeternalconn.ExternalConnEnv`.

Release note: None

ZhouXing19 commented:
Ah thanks! I rebased and re-ran `./dev gen bazel` and that did changed a bazel file. Not sure why previously it didn't work and the CI for this branch was green. 

bors r+
